File: sensor/sht20/src/main/message.c
#include "message.h"

/**
 * {"NAME":<name>,"TYPE":<type>,"VALUE":<value>,"ARG":<arg>,"TIMESTAMP":<timestamp>,"PID":<pid>,"VID":<vid>,"UUID":<UUID>}
 */

#define MESSAGE_SEND_PACKAGE "{\"NAME\":\"%s\",\"TYPE\":%d,\"VALUE\":%e,\"ARG\":%e,\"TIMESTAMP\":%d,\"PID\":%d,\"VID\":%d,\"UUID\":\"%s\"}"
#define MESSAGE_RECEIVE_PACKAGE "{\"NAME\":\"%[a-zA-Z0-9_]\",\"TYPE\":%d,\"VALUE\":%e,\"ARG\":%e,\"TIMESTAMP\":%d,\"PID\":%d,\"VID\":0x%d,\"UUID\":\"%[a-zA-Z0-9x-]\"}"

static inline void _pack(struct _message message, char *pack_buf)
{
    sprintf(pack_buf, MESSAGE_SEND_PACKAGE, message.name, message.type, message.value, message.arg0, message.timestamp, message.pid, message.vid, message.uuid);
}

void message_pack(char *name, int type, float value,  float arg, int timestamp, \
                   int pid, int vid, char *uuid, char *pack_buf)
{
    struct _message message = {
        .type = type,
        .pid = pid,
        .arg0 = arg,
        .timestamp = timestamp,
        .value = value
    };

    sprintf(message.name, "%s", name);
    sprintf(message.uuid, "%s", uuid);

    _pack(message, pack_buf);
}

void message_unpack(struct _message *message, char *raw_message)
{
    sscanf(raw_message, MESSAGE_RECEIVE_PACKAGE, message->name, &(message->type), \
                &(message->value),  &(message->arg0), &(message->timestamp), &(message->pid), \
                &(message->vid), message->uuid);
}

File: system/scripts/message.py
'''
{"NAME":<name>,"TYPE":<type>,"VALUE":<value>,"ARG":<arg>,"TIMESTAMP":<timestamp>,
"PID":<pid>,"VID":<vid>,"UUID":<UUID>}

struct _message {
    char *name;
    int type;
    float value;
    float arg0;
    int timestamp;
    int pid;
    int vid; 
    char *uuid;
};
    SYSTEM_EVENT = BIT0,
    ECHO_EVENT = BIT1,
    TIMESTAMP_EVENT = BIT2,
'''

_MESSAGE_TYPE_SYSTEM_EVENT = (1 << 0)
_MESSAGE_TYPE_ECHO_EVENT = (1 << 1)
_MESSAGE_TYPE_TIMESTAMP_EVENT = (1 << 2)

class _message:
    topic_name = "SYSTEM"
    message_type = 0
    value = 0
    arg = 0
    timestamp = 0
    pid = 0x000
    vid = 0x000
    uuid = "0x00000000-0000-1000-8000-00805F9B34FB"

    def generate():
        pass